fem/src/Lua.F90

```diff
@@ -0,0 +1,386 @@
+  character(kind=c_char, len=:), pointer :: sp
+
+  character(kind=c_char, len=:), allocatable :: s
+  type(c_ptr) :: c_s
+  integer(kind=c_int) :: slen
+
+  c_s = luaL_checkstring_c(L, n, slen)
+  call c_f_pointer(c_s, sp)
+end function
+
+subroutine lua_runfile(L, fname)
+  type(LuaState_t) :: L
+  character(kind=c_char) :: fname(*)
+  call lua_runfile_c(L%L, fname)
+end subroutine
+
+function lua_dostring(L, s, m) result(n)
+  type(LuaState_t) :: L
+  character(kind=c_char) :: s(*)
+  integer(kind=c_int) :: n, m_, load_error, pcall_error
+  integer(kind=c_int), optional :: m
+  if(.not. present(m)) then
+    m_ = -1
+  else
+    m_ = m
+  end if
+  load_error = luaL_loadstring(L%L, s)
+  call check_error(L, load_error)
+  if(load_error == 0) then
+    pcall_error = lua_pcall(L%L, 0, m_, 0)
+    call check_error(L, pcall_error)
+  end if
+  n = IOR(load_error, pcall_error)
+end function
+
+real(kind=c_double) function lua_getnumber(L, s)
+  type(LuaState_t) :: L
+  character(kind=c_char) :: s(*)
+  call lua_getfield(L%L, LUA_GLOBALSINDEX, s)
+  lua_getnumber = lua_popnumber(L)
+end function
+
+subroutine lua_poptensor(L, t)
+  type(LuaState_t) :: L
+  real(kind=c_double), intent(out) :: t(:,:)
+
+  integer :: n1, n2, i, j
+  n1 = size(t, 1)
+  n2 = size(t, 2)
+  do i = n1,1, -1
+    do j = n2,1,-1
+      t(i,j) = lua_popnumber(L)
+    end do
+  end do
+end subroutine
+
+subroutine lua_eval_f(L, fname, X, y)
+  type(LuaState_t) :: L
+  character(kind=c_char) :: fname(*)
+  real(kind=c_double), intent(in) :: X(:)
+  real(kind=c_double), intent(inout) :: Y(:)
+
+  integer :: nx, ny, i, lstat
+  nx = size(X,1)
+  ny = size(Y,1)
+  CALL lua_getfield(L%L, LUA_GLOBALSINDEX, fname)
+  do i = 1,nx
+    CALL lua_pushnumber(L%L, X(i))
+  end do
+  lstat = lua_pcall(L%L, nx, ny, 0)
+  call check_error(L, lstat)
+  if (lua_pcall(L%L, nx, ny, 0) /= 0) then
+    CALL luaL_error(L%L, "error running '"//fname(1:len(fname))//"': ")
+  end if
+  do i = ny,1,-1
+    Y(i) = lua_tonumber(L%L, -1) 
+    CALL lua_pop(L%L,1)
+  end do
+end subroutine
+
+!> Execute fname in lua state L, dont collect results from stack but expect user to collect them.
+subroutine lua_exec_fun(L, fname, nin, nout)
+  type(LuaState_t) :: L
+  character(kind=c_char), intent(in) :: fname(*)
+  integer, intent(in) :: nin, nout
+  integer :: lstat
+
+  CALL lua_getfield(L%L, LUA_GLOBALSINDEX, fname)
+  lstat = lua_pcall(L%L, nin, nout, 0)
+  call check_error(L, lstat)
+end subroutine
+
+real(kind=c_double) function lua_popnumber(L)
+  type(LuaState_t) :: L
+  lua_popnumber = lua_tonumber(L%L, -1)
+  call lua_pop(L%L,1)
+end function
+
+function lua_init() result(L)
+  type(LuaState_t) :: L
+  type(c_ptr) :: ptr
+  L%L = lua_init_c()
+  if(c_associated(L%L)) L % initialized = .true.
+end function
+
+subroutine lua_addfun(L, fun, fname)
+  type(LuaState_t) :: L
+  procedure(luafun), pointer :: fun
+  type(c_funptr) :: c_fun
+  character(kind=c_char):: fname
+
+  c_fun = transfer(fun, c_fun)
```

Comment:
I don't see the point in using `transfer` in this case. In addition, such a line seems to crash ifort. It would be better to replace the line with `c_fun = c_funloc(fun)` as it achieves the same functionality via standard Fortran methods.